ChiModules/KEigenvalueSolver/IterativeMethods/power_iteration.cc

```diff
@@ -25,161 +28,128 @@ using namespace LinearBoltzmann;
+  ScopedCopySTLvectors(groupset, phi_prev_local, phi_old_local);
 
+  //======================================== Start power iterations
+  double F_prev = 1.0;
+  double k_eff_prev = 1.0;
+  int nit = 0;      //number of iterations
+  bool converged = false;
   while (nit < options.max_iterations)
   {
-    chi_log.Log(LOG_0VERBOSE_2)
-      << "\n********** Starting source iterations";
 
+    //============================== Clear source moments
+    q_moments_local.assign(q_moments_local.size(), 0.0);
+
+    //============================== Set the fission source
+    SetKSource(groupset, q_moments_local,
+               APPLY_AGS_FISSION_SOURCE |
+               APPLY_WGS_FISSION_SOURCE);
 
-    // ----- Start inner source iterations
-    double pw_change_prev = 1.0;
-    bool si_converged = false;
-    for (int si_nit=0; si_nit<groupset.max_iterations; si_nit++)
+    //============================== Converge the scattering source with
+    //                               a fixed fission source
+    if (groupset.iterative_method == IterativeMethod::CLASSICRICHARDSON)
     {
-      // ----- Set source and sweep
-      SetKSource(groupset, APPLY_MATERIAL_SOURCE |
-                           APPLY_AGS_SCATTER_SOURCE | APPLY_WGS_SCATTER_SOURCE |
-                           APPLY_FISSION_SOURCE);
-      groupset.angle_agg.ZeroOutgoingDelayedPsi();
-      phi_new_local.assign(phi_new_local.size(),0.0);
-      SweepScheduler.Sweep();
-
-      // ----- Compute convergence parameters
-      double pw_change = ComputePiecewiseChange(groupset);
-      ScopedCopySTLvectors(groupset, phi_new_local, phi_old_local);
-      double rho = sqrt(pw_change/pw_change_prev);
-      pw_change_prev = pw_change;
-      nit += 1;
-
-      if (si_nit==0) rho = 0.0;
-      if (pw_change<std::max(groupset.residual_tolerance*rho,1.0e-10))
-        si_converged = true;
-
-      // ----- Print iteration information
-      std::string offset = "    ";
-      std::stringstream si_iter_info;
-      si_iter_info
-        << chi_program_timer.GetTimeString() << " "
-        << offset
-        << "WGS groups ["
-        << groupset.groups.front().id
-        << "-"
-        << groupset.groups.back().id
-        << "]"
-        << " Source Iteration " << std::setw(5) << si_nit
-        << " Point-wise change " << std::setw(14) << pw_change;
-        
-      if (si_converged)
-        si_iter_info << " CONVERGED\n";
-      chi_log.Log(LOG_0VERBOSE_1) << si_iter_info.str();
-
-      if (si_converged) {
-        chi_log.Log(LOG_0VERBOSE_1)
-            << "\nSource iterations converged in "
-            << si_nit << " iteratrions.\n";
-        break;
-      }
-    }//for source iterations
-
-    if (!si_converged) {
-      chi_log.Log(LOG_ALLVERBOSE_1)
-          << "\n!!!WARNING!!! "
-             "Source iterations did not converge.\n";
+      ClassicRichardson(groupset, 0, sweep_scheduler,
+                        APPLY_WGS_SCATTER_SOURCE | APPLY_AGS_SCATTER_SOURCE,
```

Comment:
I think that the `APPLY_MATERIAL_SOURCE` flag needs to be added here. Even though there isn't any fixed source, per #156 I think we won't have reflecting boundaries without a `APPLY_MATERIAL_SOURCE` flag

Reply:
@Naktakala : at some point, would it be possible (if it does not yet exist) to have a short document (even an .md) describing the various APPLY_xxxx, their obvious functions but also their maybe-less-obvious functions as well?

Reply:
I was actually alarmed by this @jthano because I initially thought it is not correct. The `APPLY_MATERIAL_SOURCE` flag is used with a sweep-chunk setting to determine whether incident boundary conditions are to be applied during a sweep. However, if the boundary is reflecting then this flag needs to be ignored. Surprise: It's not being ignored as can be seen here

```c++
double* chi_mesh::sweep_management::AngleSet::
PsiBndry(uint64_t bndry_map,
         int angle_num,
         uint64_t cell_local_id,
         int face_num,
         int fi,
         int g,
         int gs_ss_begin,
         bool surface_source_active)
{
 if (not surface_source_active)
    return &ref_boundaries[bndry_map]->zero_boundary_flux[g];

  if (ref_boundaries[bndry_map]->IsReflecting())
    return ref_boundaries[bndry_map]->HeterogenousPsiIncoming(
      angle_num, cell_local_id, face_num, fi, gs_ss_begin);

  return &ref_boundaries[bndry_map]->boundary_flux[g];
}
```

Now ... I am aggressively confident that somewhere in the past I changed this function and induced this bug. I shall fix it and the underlying lack of a regression which should've pointed this out as it only affects GMRES.

Reply:
I have my C5G7 benchmark up and running, but I'm not getting the correct eigenvalue. This issue with reflecting boundaries is one thing I ran into while trying to get my benchmark working. I actually thought that #156 was specifically to fix issues related to reflecting boundaries.

Reply:
This issue is addressed in #159 and further brought up in #158.